lib/timers.js

```diff
@@ -440,13 +452,20 @@ function rearm(timer) {
 
 
 const clearTimeout = exports.clearTimeout = function(timer) {
-  if (timer && (timer[kOnTimeout] || timer._onTimeout)) {
+  if (!timer) return;
+
+  if (timer[kOnTimeout] || timer._onTimeout) {
+    if (typeof timer._onTimeout !== 'function') {
+      promiseResolve(timer._onTimeout, undefined);
+    }
     timer[kOnTimeout] = timer._onTimeout = null;
     if (timer instanceof Timeout) {
       timer.close(); // for after === 0
     } else {
       unenroll(timer);
     }
+  } else if (isPromise(timer)) {
```

Comment:
> First of all, attaching properties to native objects seems like a bad idea from performance perspective.

That’s true, but it’s opt-in, right?

> This approach suffers from the fact that it's not propagated

What do you mean by propagated?

But yeah, I would be okay with dropping `clear*` support for these, at least for the time being.